lib/AST/ASTMangler.cpp

```diff
@@ -2293,8 +2293,17 @@ void ASTMangler::appendFunctionInputType(
   default:
     bool isFirstParam = true;
     for (auto &param : params) {
-      appendTypeListElement(Identifier(), param.getPlainType(),
-                            param.getParameterFlags(), forDecl);
+      // We need to include the argument label for the enum case in the
+      // mangling as well, to avoid collisions when we have enum cases
+      // that share the same base name, but have different labeled
+      // arguments.
+      if (isa<EnumElementDecl>(forDecl) && param.hasLabel()) {
```

Comment:
This feels like a hack because clearly we don't have this problem for normal FuncDecls. Can you make EnumElementDecls take the FuncDecl path instead?

Reply:
Hmm, let me investigate further.

Given we have a type like `(EnumMetaType) -> (AssocValues) -> Enum`,  you end up with a code path like: `appendEntity -> appendDeclType -> appendFunction -> appendFunctionSignature -> appendFunctionResultType -> appendType -> appendFunctionType -> appendFunctionSignature`.

So the mangler calls `appendFunctionResultType ` with the result type (which is `(AssocValues) -> Enum`) which leads us to `appendFunctionInputType` where we append the `(AssocValues)` but we drop labels.